pkg/audit/stats_reporter.go

```diff
@@ -2,111 +2,115 @@ package audit
-	views := []*view.View{
-		{
-			Name:        violationsMetricName,
-			Measure:     violationsM,
-			Aggregation: view.LastValue(),
-			TagKeys:     []tag.Key{enforcementActionKey},
-		},
-		{
-			Name:        auditDurationMetricName,
-			Measure:     auditDurationM,
-			Aggregation: view.Distribution(1*60, 3*60, 5*60, 10*60, 15*60, 20*60, 40*60, 80*60, 160*60, 320*60),
-		},
-		{
-			Name:        lastRunStartTimeMetricName,
-			Measure:     lastRunStartTimeM,
-			Aggregation: view.LastValue(),
-		},
-		{
-			Name:        lastRunEndTimeMetricName,
-			Measure:     lastRunEndTimeM,
-			Aggregation: view.LastValue(),
-		},
+	auditDurationM, err = meter.Float64Histogram(
+		auditDurationMetricName,
+		metric.WithDescription("Latency of audit operation in seconds"))
+	if err != nil {
+		panic(err)
 	}
-	return view.Register(views...)
-}
 
-func (r *reporter) reportTotalViolations(enforcementAction util.EnforcementAction, v int64) error {
-	ctx, err := tag.New(
-		context.Background(),
-		tag.Insert(enforcementActionKey, string(enforcementAction)))
+	lastRunStartTimeM, err = meter.Float64ObservableGauge(
+		lastRunStartTimeMetricName,
+		metric.WithDescription("Timestamp of last audit run starting time"),
+	)
 	if err != nil {
-		return err
+		panic(err)
 	}
 
-	return r.report(ctx, violationsM.M(v))
-}
-
-func (r *reporter) reportLatency(d time.Duration) error {
-	ctx, err := tag.New(context.Background())
+	lastRunEndTimeM, err = meter.Float64ObservableGauge(
+		lastRunEndTimeMetricName,
+		metric.WithDescription("Timestamp of last audit run ending time"),
+	)
 	if err != nil {
-		return err
+		panic(err)
 	}
 
-	return r.report(ctx, auditDurationM.M(d.Seconds()))
+	view.Register(sdkmetric.NewView(
+		sdkmetric.Instrument{Name: auditDurationMetricName},
+		sdkmetric.Stream{
+			Aggregation: sdkmetric.AggregationExplicitBucketHistogram{
+				Boundaries: []float64{1 * 60, 3 * 60, 5 * 60, 10 * 60, 15 * 60, 20 * 60, 40 * 60, 80 * 60, 160 * 60, 320 * 60},
+			},
+		},
+	))
 }
 
-func (r *reporter) reportRunStart(t time.Time) error {
-	ctx, err := tag.New(context.Background())
-	if err != nil {
-		return err
+func (r *reporter) registerCallback() error {
+	_, err1 := meter.RegisterCallback(r.observeTotalViolations, violationsM)
+	_, err2 := meter.RegisterCallback(r.observeRunEnd, lastRunEndTimeM)
+	_, err3 := meter.RegisterCallback(r.observeRunStart, lastRunStartTimeM)
+	return errors.Join(err1, err2, err3)
+}
+
+func (r *reporter) observeTotalViolations(_ context.Context, o metric.Observer) error {
+	for k, v := range r.totalViolationsPerEnforcementAction {
+		o.ObserveInt64(violationsM, v, metric.WithAttributes(attribute.String(enforcementActionKey, string(k))))
 	}
+	return nil
+}
 
-	val := float64(t.Unix())
-	return metrics.Record(ctx, lastRunStartTimeM.M(val))
+func (r *reporter) reportLatency(d time.Duration) error {
+	auditDurationM.Record(context.Background(), d.Seconds())
+	return nil
 }
 
-func (r *reporter) reportRunEnd(t time.Time) error {
```

Comment:
Per comment in `audit/manager.go`, this function (`reportRunEnd`) should be what modifies r.endTime. Also, we should either set a mutex or use an atomic variable to avoid concurrency issues.

The same is true for any other `reportX` function that has been deleted here and in any other `_reporter.go` file.